pep-0612.rst

```diff
@@ -128,169 +156,346 @@ will defer the standardization of the semantics of those options to a later PEP.
-contains a ``ParameterSpecification`` variable are analogous to those around
-evaluating ones with ``TypeVar``\ s. 
+contains a ``ParamSpec`` variable are analogous to those around
+evaluating ones with ``TypeVar``\ s.
 
 .. code-block::
 
-   def foo(
-       x: typing.Callable[TParams, int]
-   ) -> typing.Callable[TParams, str]: ...
+   def foo(x: Callable[P, int]) -> Callable[P, str]: ...
+
    def bar(a: str, b: bool) -> int: ...
-   f = foo(bar) # f should be inferred to have the same signature as bar, 
+
+   f = foo(bar) # f should be inferred to have the same signature as bar,
                 # but returning str
-   f("A", True) # Accepted
-   f(a="A", b=True) # Accepted
-   f("A", "A") # Rejected
+
+   f("A", True)               # Accepted
+   f(a="A", b=True)           # Accepted
+   f("A", "A")                # Rejected
+
+   expects_str(f("A", True))  # Accepted
+   expects_int(f("A", True))  # Rejected
 
 Just as with traditional ``TypeVars``\ , a user may include the same
-``ParameterSpecification`` multiple times in the arguments of the same function,
+``ParamSpec`` multiple times in the arguments of the same function,
 to indicate a dependency between multiple arguments.  In these cases a type
 checker may choose to solve to a common behavioral supertype (i.e. a set of
 parameters for which all of the valid calls are valid in both of the subtypes),
 but is not obligated to do so.
 
 .. code-block::
 
-   def foo(
-       x: typing.Callable[TParams, int], y: typing.Callable[TParams, int]
-   ) -> typing.Callable[TParams, bool]: ...
+   P = ParamSpec("P")
+
+   def foo(x: Callable[P, int], y: Callable[P, int]) -> Callable[P, bool]: ...
 
    def x_int_y_str(x: int, y: str) -> int: ...
    def y_int_x_str(y: int, x: str) -> int: ...
-   foo(x_int_y_str, x_int_y_str) # Must return (x: int, y: str) -> int
-   foo(x_int_y_str, y_int_x_str) # Could return (__a: int, __b: str) -> int 
-                                 # This works because both callables have types 
-                                 # that are behavioral subtypes of 
-                                 # Callable[[int, str], int]
 
-   def keyword_only_x(*, x: int) -> int: ...
-   def keyword_only_y(*, y: int) -> int: ...
-   foo(keyword_only_x, keyword_only_y) # Must be rejected
+   foo(x_int_y_str, x_int_y_str) # Should return (x: int, y: str) -> bool
+
+   foo(x_int_y_str, y_int_x_str) # Could return (__a: int, __b: str) -> bool
+                                 # This works because both callables have types
+                                 # that are behavioral subtypes of
+                                 # Callable[[int, str], object]
+
+   def keyword_only_x( *, x: int) -> int: ...
+   def keyword_only_y( *, y: int) -> int: ...
+   foo(keyword_only_x, keyword_only_y) # Rejected
+
+The semantics of ``Concatenate[X, Y, P]`` are that it represents the parameters
+represented by ``P`` with two positional-only parameters prepended.  This means
```

Comment:
nit: link to PEP570 for positional-only arg?

Reply:
This is actually a PEP 484 positional only notation (https://www.python.org/dev/peps/pep-0484/#positional-only-arguments), and generally I think it may actually read poorer with a reference to other materials